src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -182,191 +184,176 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
-
-	return recoveredTransactionCount;
-}
-
-
-/*
- * NameListDifference returns the difference between the bag of
- * names in nameList and subtractList. Both are assumed to be
- * sorted. We cannot use list_difference_ptr here since we need
- * to compare the actual strings.
- */
-static List *
-NameListDifference(List *nameList, List *subtractList)
-{
-	List *differenceList = NIL;
-	ListCell *nameCell = NULL;
-
-	int subtractIndex = 0;
-	int subtractCount = list_length(subtractList);
-	char **subtractArray = (char **) PointerArrayFromList(subtractList);
-
-	foreach(nameCell, nameList)
-	{
-		char *baseName = (char *) lfirst(nameCell);
-
-		bool nameFound = FindMatchingName(subtractArray, subtractCount,
-										  baseName, &subtractIndex);
-
-		if (!nameFound)
+		/*
+		 * Remove the transaction from the pending list such that only transactions
+		 * that need to be aborted remain at the end.
+		 */
+		hash_search(pendingTransactionSet, transactionName, HASH_REMOVE,
+					&isTransactionPending);
+
+		/*
+		 * Check if we found a prepared transaction the second time around.
+		 *
+		 * If a transaction committed just before we got our pg_dist_transaction
+		 * snapshot then it may have left prepared transactions that should
+		 * be committed, but are not in pendingTransactionSet. Therefore we
+		 * check the prepared transaction a second time. Otherwise, we would
+		 * incorrectly delete them without comitting the prepared transaction.
+		 */
+		hash_search(recheckTransactionSet, transactionName, HASH_FIND,
+					&foundNewPreparedTransaction);
+
+		if (isTransactionPending || foundNewPreparedTransaction)
 		{
 			/*
-			 * baseName is not in subtractArray and thus included
-			 * in the difference.
+			 * The transaction was committed, but the prepared transaction still exists
+			 * on the worker. Try committing it.
 			 */
-			differenceList = lappend(differenceList, baseName);
+			bool shouldCommit = true;
+			bool commitSucceeded = RecoverPreparedTransactionOnWorker(connection,
+																	  transactionName,
+																	  shouldCommit);
+			if (!commitSucceeded)
+			{
+				/*
+				 * Failed to commit on the current worker. Stop without throwing
+				 * an error to allow recover_prepared_transactions to continue with
+				 * other workers.
+				 */
+				recoveryFailed = true;
+				break;
+			}
+
+			recoveredTransactionCount++;
 		}
-	}
-
-	pfree(subtractArray);
-
-	return differenceList;
-}
-
-
-/*
- * CompareNames compares names using strncmp. Its signature allows it to
- * be used in qsort.
- */
-static int
-CompareNames(const void *leftPointer, const void *rightPointer)
-{
-	const char *leftString = *((char **) leftPointer);
-	const char *rightString = *((char **) rightPointer);
-
-	int nameCompare = strncmp(leftString, rightString, NAMEDATALEN);
-
-	return nameCompare;
-}
 
+		/* we've successfully committed the prepared transaction, remove the record */
+		simple_heap_delete(pgDistTransaction, &heapTuple->t_self);
```

Comment:
Nothing actionable, just noting for future reference: The perf output follows:

```
      - PostgresMain                                                                                                                                                                                                                  ▒
         - 77.59% exec_simple_query                                                                                                                                                                                                   ▒
              PortalRun                                                                                                                                                                                                               ▒
              PortalRunSelect                                                                                                                                                                                                         ▒
              ExecutorRun                                                                                                                                                                                                             ▒
              standard_ExecutorRun                                                                                                                                                                                                    ▒
              ExecutePlan                                                                                                                                                                                                             ▒
              ExecProcNodeFirst                                                                                                                                                                                                       ▒
              ExecResult                                                                                                                                                                                                              ▒
              ExecInterpExpr                                                                                                                                                                                                          ▒
              recover_prepared_transactions                                                                                                                                                                                           ▒
            - RecoverPreparedTransactions                                                                                                                                                                                             ▒
               - 77.12% RecoverWorkerTransactions                                                                                                                                                                                     ▒
                  + 31.20% simple_heap_delete                                                                                                                                                                                         ▒
                  + 29.06% systable_getnext                                                                                                                                                                                           ▒
                  + 8.08% hash_search                                                                                                                                                                                                 ▒
                  + 3.38% text_to_cstring                                                                                                                                                                                             ▒
                  + 2.64% IsTransactionInProgress                                                                                                                                                                                     ▒
                    0.92% __memcpy_ssse3                                                                                                                                                                                              ▒
                    0.56% MemoryContextDelete 
```

So, we're mostly bottlenecked on iterating and deleting from `pg_dist_transaction` which seems reasonable to me.

The largest test I've done was with 35M rows in `pg_dist_transaction`, which took ` 1064 secs` without blocking the new 2PCs on `1+2 c4.xlarge` cluster. The memory usage increased to `1GB`, but, seems OK to me.

With a smaller number of entries, the performance is obviously much better. (512000 entries in `46 secs`)